pageserver/src/storage_sync/index.rs

```diff
@@ -262,19 +262,88 @@ impl RemoteTimeline {
 
 /// Part of the remote index, corresponding to a certain timeline.
 /// Contains the data about all files in the timeline, present remotely and its metadata.
-#[serde_as]
-#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
-pub struct IndexPart {
-    timeline_layers: HashSet<RelativePath>,
-    /// Currently is not really used in pageserver,
-    /// present to manually keep track of the layer files that pageserver might never retrieve.
-    ///
-    /// Such "holes" might appear if any upload task was evicted on an error threshold:
-    /// the this layer will only be rescheduled for upload on pageserver restart.
-    missing_layers: HashSet<RelativePath>,
-    #[serde_as(as = "DisplayFromStr")]
-    disk_consistent_lsn: Lsn,
-    metadata_bytes: Vec<u8>,
+///
+/// Important to note: none of the payloads can contain "version" field due to it's usage in the
+/// envelope.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum IndexPart {
+    /// With or without a "version" field. Note, the serialized form will not transparently upgrade the type.
+    V1 { inner: IndexPartV1, versioned: bool },
+    // Later versions can be added as:
+    // V2(IndexPartV2),
+}
+
+/// The optional version identifier of [`IndexPart`] (de)serialization.
+#[derive(Serialize, Deserialize)]
+enum IndexPartVersion {
+    /// First version as of 2022-10-04 a50e46f5, and earlier.
+    #[serde(rename = "1")]
+    V1,
+    // later versions will need a new tag, like:
+    // #[serde(rename = "2")]
+    // V2,
+}
+
+/// Non-derived implementation that sometimes puts the version field in the document.
+impl Serialize for IndexPart {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        use IndexPart::*;
+
+        /// The envelope to produce the serialized version of optionally internally tagged document.
+        #[derive(Serialize)]
+        struct Versioned<'a, I: Serialize> {
+            version: IndexPartVersion,
+            #[serde(flatten)]
+            inner: &'a I,
+        }
+
+        let envelope = match self {
+            // V2(inner) => Versioned {
+            //     version: IndexPartVersion::V2,
+            //     inner,
+            // },
+            V1 { inner, versioned } if *versioned => Versioned {
```

Comment:
I'm not sure what `versioned` is supposed to be? An adapter to mitigate the absence of version in the existing format?

From what I understand, the format does not change, but gets the version in the new approach, so it might make sense to serialize it with V1 as basic version, when it's absent.

Reply:
Forgot to post this, but I was thinking of not doing automatic version promotion, so that `&serde_json::to_string(serde_json::from_str(&without_version)?)? == without_version`, but otherwise it's not needed.

Reply:
Yeah, I think we need to discuss that a bit more, are there any downsides with doing so?